### ISSUE TITLE
Why the chunk replication speed slows down after a while? 

### ISSUE THREAD

**Issue by hradec** [lizardfs/lizardfs#788]
I've notice this behavior in 2 systems I've setup. 
After changing goals for files, LFS starts replication and goes fast and steady... as soon as it starts to get to the end, it starts to slow down until it stays literally CRAWLING to a halt, compared to when it started. 

The most frustrating part of this is, because of this slow down, the very last chunks needed to replicate take weeks, which is very concerning since a HD could fail and we would loose data. 

this is a screenshot of the behavior... a week of replication after a few changes in goal for all files in the pool:
![image](https://user-images.githubusercontent.com/806743/48986645-8b8bf500-f0cc-11e8-8a0a-3a9e39ca4374.png)
I'm feeding grafana with collected data from lizardfs-admin stored in telegraf. This gives me a way better visualization of what LFS is doing in the long term.

as you can see, at the begining of the week, there was a lot of chunk decrease due to replication happening. Most of the replication was being done to a new chunk I've added to the system (11 chunks total).

There was about 250.000 chunks to be replicated at the begining of the week... in about 4 days, this number went down to 62.000 (almost 50.000 a day!)... since then, it's being 3 days and I still have 54.000 to go, and the rate had being slowing down exponentially!!! 

this is the graph for the last 24 hours:
![image](https://user-images.githubusercontent.com/806743/48986739-62b82f80-f0cd-11e8-99ae-a82cee199f73.png)

As you can see, since today around 2pm, it's doing almost NOTHING, and I still have 54.000 chunks to go!! 
![image](https://user-images.githubusercontent.com/806743/48986750-87140c00-f0cd-11e8-811a-2d48f7b74d7a.png)

It's actually very frustrating to see this slow down happening so often. I've actually changed the goals last week just to try to force the pool to FINISH replicating, since it's always doing this slow down at the end... This is the 4th time I change goals in the hopes of seeing the replication finishing and finally known that my data is safe! 

but it NEVER ends... It's being 5 months that the pool NEVER finishes replicating and my data is at risk!! 

Today, if I take 1 chunk offline, I always end up with missing chunks... no matter what chunk I take offline! 
The goals are ec(6,2), so in theory, I could take 2 chunks out without loosing anything. But since it never finishes replicating, I have loads of data at risk! 

Yesterday I've set CHUNKS_WRITE_REP_LIMIT=100 and CHUNKS_READ_REP_LIMIT=200, but it had NO EFFECT WHATSOEVER, as you can see on the graphs above... Right now, the replication is almost halting, with those to values!! 

so if any of my chunks really die today, I'll loose data for sure because of this slow down... 

please help!! I need to be able to forcibly end the replication!!!  

PS: as another example, this is another LFS pool that I'm running for more than a year (I think it's almost 2 years actually): 

![image](https://user-images.githubusercontent.com/806743/48986998-ac097e80-f0cf-11e8-9c3e-520cfcb1b09d.png)

Apart from the cp(1) goals, you can see that I have this nexenta_ec6_2  ec(6,2) goal  which still need replicating, and they have being so slow that they replicate at a rate of 1 chunk per DAY!!! 
it's being 4 MONTHS to get from around 5.000 chunks down to 1146 (1 copy replicate), and it's still going.. 

As you can see, I do have missing chunks in this pool because since I started it, I did have failed hard drives, and and indeed lost chunks thanks to this slow down behavior!!! 

I think this is a very serious issue in LizardFS, one that I don't see people giving too much attention. I did see some people complaining, but the answer is always to tweak CHUNKS_* parameters in the master to trade-of replication speed by user access latency... But I REALLY don't think this is the solution! 

Theres something seriously wrong in the way LizardFS goes about the replication tasks, and it really looses momentum during replication of large number of chuncks, which puts the data in risk for months and months... and there's absolutely NOTHING we can do but wait and cross our fingers right now!! 

I'm going back to evaluating other cluster filesystem out there right now because I really need to sleep at night without fearing that I'm going too loose my data. 

very sad indeed... I really like LFS... but after 2 years, I still can't trust it because it never finishes!

**Comment by njhurst** [comment]
I think you don't have enough chunkservers, also, having chunkservers at 100% gives me the same stagnating behaviour.  Try splitting your two chunkservers with 50% usage - just make a new directory and start a new server pointed at that directory.  I believe, based on observation, that to go from 6,3 to 6,2 you actually need 17 chunkservers.

If you find a better cluster system do come back and share!

**Comment by Zorlin** [comment]
As a rule of thumb I don't like to push any individual chunkserver past ~80% utilization. I worry at 85%, and then I panic at 90%.

**Comment by hradec** [comment]
well, I didn't "let" it fill up! I actually had no choice, since LizardFS keeps feeling up those 4 chunkservers no matter what!!  Even with a 4TB chunkserver with only 50% free, it goes on filling up the 500gb/1TB/2TB chunkservers like that!! 

it was LizardFS who choose to fill up chunkservers this way!!! Not me! (all my chunkservers have the same label!)

It seems it keeps filling up a chunkserver until it hits 99.NN or even 100% (like my /RAID/chunk over there), and it NEVER starts re-balancing!! (and yes, I do have CHUNKS_REBALANCING_BETWEEN_LABELS=1)

the thing is... I have 11 chunkservers, so for a ec(6,2) which needs 8 servers to accomodate each of the 6+2 parts, 11 chunks should be 3 spares!  (192.168.10.103 is a chunkserver with 2 disks, actually, and booth disks are 50% and 75% full only!)
![image](https://user-images.githubusercontent.com/806743/48995967-0ec63e80-f0ff-11e8-850a-c40be3edf711.png)

The first time I changed goals was from cp(2) to ec(6,2)... after that, I have being switching from ec(6,2) to another ec(6,2) just to force it to keep going due to the slow down behaviour in the first place!

The problem here is that at the start of the week (when I had 200.000 chunks to go), my servers where beetwen 70% and 40%, and they where stagnated... really crawling replication, no rebalancing! So I've switched the goals and It kept going! (at the cost of wasting space until it finishes and starts cleaning up!)

But It doesn't make sense... shouldn't the master starts to rebalance when it figures its going to run out of space? 

It's actually ridiculous that there's no way to force it to rebalance, so the only solution for a problem created by LFS is to keep adding servers/disks! 

I really want to avoid creating a new folder with a new chunkserver on a disk that already is in the pool, since this would actually create a hidden deficit in redundancy!

What's the point of having ec(6,2), but having 2 parts on the same disk (because of different chunkservers using the same disk)... if 2 disks fail and one of then is said disk, I loose 3 parts and my data is lost!

The only solution here is to keep adding new disks until replication finishes, and then it will re-balances?? or I have to remove the added disk and wait it to replicate again?!?!

It's too weird... Why LFS doesn't Rebalances instead of crawling? 
It should just flag that we don't have space on a few chunks and prioritize rebalancing, so it can keep going!! Or at least we should be able to FORCE it to rebalance, so we don't have to panic when LFS puts 99% data on a disk, and leaves 3 more at 50%!

thanks for the help, guys! I really appreciate... it seems the only solution here is buy more disks and hardware to finish this, sleep well, and start migrating to something more reliable! 

Apart from this, I'm pretty happy with LFS! It's amazingly simple and easy to setup!
But I can't keep living with this HUGE doubt on the back of my head everyday, like I have being for the past few months... 

again, thanks lots... 
-H

**Comment by hradec** [comment]
![image](https://user-images.githubusercontent.com/806743/48996705-f0157700-f101-11e8-8e36-37928baa2511.png)

**Comment by hradec** [comment]
> As I said above, it seems that lizard is unwilling to share chunkservers between two partial copies of a goal change. 

That's weird... in theory, since the goal change is the same number of parts, it should just be a matter of switching the goal ID on those parts! (at least thats what happens with CP(2) when I first tested)

There's no point in re-construct the ec(6,2) for what is already done, and in fact, when I switch goals this way, it seems to replicate insanely fast until it gets to the original number of parts to replicate that I had before the change! 

For example, when I switched from one ec(6,2) ID to those 8 ec(6,2) IDs, I had 200.000 chunks to go... when I switched, I got millions of chunks to replicate, but in a matter of hours, I had the same 200.000 chunks to go and then it kept going from there for about 3 days to 62.000, where it started to slow down  to the crawling that I have today. 

So it does seem to just switch the goal ID for what is already done, and this goal change indeed fixed the crawling... but at the cost of filling up disks! 

I just don't get it why it doesn't just prioritizes re-balances to free up space! 

>The solution is to partition a drive into multiple chunkservers. Yes, you will need to deprecate the extras once the rebalance is fixed.

that's actually I question... if I do that, and then deprecate the extras, wouldn't I run into the same replication problem again?!!?! (like it happened when I added those 2 extra disks?)

actually, one of those 12 disks is already an extra that I added to finish... I already have 1TB that I need to remove once it finishes! LOL

I truly believe that this behavior is not correct... It does look like a bug that a simple temporary replication prioritization would fix for good!!  Just make it the master force the pool to rebalance, get free space on the chunks as fast as it can, and keep going! 

just to update... this is the last 3 hours: 
![image](https://user-images.githubusercontent.com/806743/48997783-f148a300-f105-11e8-8bd5-3c8054157417.png)

I found some trash data on that chunk that was 100% full, deleted and restarted it... so now it's 99.42% full.. no change in replication rate!
![image](https://user-images.githubusercontent.com/806743/48997844-1ccb8d80-f106-11e8-8ed2-3f1d4a069926.png)

**Comment by BloodBlight** [comment]
This may be completely unrelated, but with an ec of 6,2 and your largest disk being 3.4TBs, you would need about 27.2 TBs of raw disk space to be able to successfully level your data.  But you only have about 22.  This SHOULDN'T be an issue (as you have 8+ chunk servers with free space), but could be causing some sort of a math/placement issue.

Maybe try setting your ACCEPTABLE_DIFFERENCE really low (like 0.01)?  

I seem to recall getting one of my install "stuck" during a goal change when I had a chunk server run out of space.  When I did that, I seem to recall it didn't want to give the space back.

Trying to remember exactly how I fixed that...  I think it was a master server restart but...  I had a script at one point that would do a file copy and swap of the whole structure, but...

**Comment by hradec** [comment]
> This may be completely unrelated, but with an ec of 6,2 and your largest disk being 3.4TBs

That's interesting... I didn't known about this and does make sense! Basically I would need (6+2)*3.4 total disk space to finish up! 





> This SHOULDN'T be an issue (as you have 8+ chunk servers with free space), but could be causing some sort of a math/placement issue.

hummm... 




> Maybe try setting your ACCEPTABLE_DIFFERENCE really low (like 0.01)?

bummer... I actually did this ACCEPTABLE_DIFFERENCE=0.01 about 4 days ago... didn't help! :(


> Trying to remember exactly how I fixed that... I think it was a master server restart but... I had a script at one point that would do a file copy and swap of the whole structure, but...

I was actually thinking about that... copy a whole bunch of files to "brute force" it to replicate!

I've actually did this once with ZFS to remove de-duplication from a pool that was making a server run out of memory.
As turning de-dup off in zfs won't automatically replicate already de-duplicated data, you have to copy it after turning de-dup off and remove the original. Only after you that, zfs stop eating memory so much! (at least back then!)

Thanks @BloodBlight... really appreciate the feedback!! I'll definitely keep that (6+2)*(biggest drive) formula in mid for future reference!! :+1:

**Comment by hradec** [comment]
just another quick update: 
![image](https://user-images.githubusercontent.com/806743/49031729-a491b600-f15f-11e8-9950-ac42db2de8e2.png)
It seems like it is finally re-balancing this morning... at a rate of about 40 chunks in the last 3 hours!!!! 8[

It is doing something... but very slowly... 

this is the disk max read/write time heatmap of the last 3 hours:
![image](https://user-images.githubusercontent.com/806743/49031591-41a01f00-f15f-11e8-8e3a-604a63ad51de.png)
There was a spike in reading at 8am today, but it didn't impact writing at all, so not sure what it was!

**Comment by BloodBlight** [comment]
If you do that, be sure to set the trash time to zero on the file before
you delete it.

Or you will end up in a worse place.

**Comment by eleaner** [comment]
@hradec what is the visualisation tool you're using?

**Comment by BloodBlight** [comment]
Setting CHUNKS_LOOP_PERIOD really low (like 100) temporarily may help speed things up.

**Comment by BloodBlight** [comment]
Sorry to spam this, but I am wondering why you have so many goals with the same settings?  Not that I think this hurts anything, but it can't be helping ether.

And back to the EC 6,2 thing...  The more I think about it, the more I that is your issue.  :/  Maybe switch to 4,2 (14.4 usable for your config).  Maybe try it on a small subset and see if it suddenly jumps into replication?  Also, I think on my smaller setup a CHUNKS_WRITE_REP_LIMIT of 12 was my point of diminishing returns.  With it set to 100 you might cause other issues.

**Comment by hradec** [comment]
>  @BloodBlight  If you do that, be sure to set the trash time to zero on the file before you delete it. Or you will end up in a worse place.
hooo..good one!  thanks!!!! 

> @eleaner what is the visualisation tool you're using?
Grafana with telegraf... I can share the telegraf script/config and the grafana json if you want! 

> @BloodBlight  Sorry to spam this, but I am wondering why you have so many goals with the same settings? Not that I think this hurts anything, but it can't be helping ether.
Before I was using just one ec(6,2) goal... but this last time I decided to split the goal names to see if it would stop slowing down, and also I could monitor what folder in my pool was getting same!

Basically, the goals suffix is the name of the folder in the root path of the pool, like /media, /miscStuff, /mirror, etc...

It actually was a good choice, since now I at least known the folders /media and /airport1TB are fairly safe, with just 2 and 5 chunks to go... if 3 disks fails today, those 2 folder are the only ones that I wont loose 99.99% of the data! 



> @BloodBlight And back to the EC 6,2 thing... The more I think about it, the more I that is your issue. :/ Maybe switch to 4,2 (14.4 usable for your config). Maybe try it on a small subset and see if it suddenly jumps into replication? Also, I think on my smaller setup a CHUNKS_WRITE_REP_LIMIT of 12 was my point of diminishing returns. With it set to 100 you might cause other issues.

The main reason for the 6,2 choice was to save space! 6,2 means the data will use only 33% more space for redudancy, with 2 disks to fail! 

4,2 means 50% more space... I could try to get the MiscStuff goal and switch it to 4,2 just to see what happens... but I need to see how much space the MiscStuff folder is taking right now.

Since re-balancing is so damn slow, I still have 5 disks above 90%, where 4 disks are 99%!

I really wish I could just brute force a re-balancing... even if I had to bring the pool down for that, but at least it would re-balance!! (sorry, but it's being hard to get over this... LOL.. I'll shut up now! LOL)

**Comment by hradec** [comment]
> @BloodBlight  Setting CHUNKS_LOOP_PERIOD really low (like 100) temporarily may help speed things up.

I'll try that! thanks!

**Comment by hradec** [comment]
btw, this is source of frustration... I keep seeing this on my mfsmaster logs all the time, even before the slow down:
![image](https://user-images.githubusercontent.com/806743/49038878-1293a880-f173-11e8-8ab2-a9b009131d66.png)

and the log of one of those chunks, is this: 
![image](https://user-images.githubusercontent.com/806743/49039103-a1a0c080-f173-11e8-9efc-67c6ee026085.png)


Funny thing is... I'm running this mfsmaster and the chunks on the same server, in LXC containers (proxmox)! So the network communication is the fastest it can be between then, since the packets never go out of the machine! 

But still, it seems to be having communication timeouts... and magical "unknown" errors! 

I've tried to look for any references of this "replication status: Unknown LizardFS error", but the only comments I've found here was "if you see that message, means the master caught an error and fixed it, so you don't have to worry about it" ! 8O

**Comment by BloodBlight** [comment]
YEP!  Totally understand.  :) 

I also get "Unknown Errors" even on a single node.  It is a bit of a catch all unless you enable verbose debugging (as far as this user can tell).

But I ran the numbers with your disk sizes, and even with EC of 6,2 you are only getting 800GBs more space than EC 4,2 (as about 2.5TBs are left unused).

Not saying you shouldn't, especially if you plan on adding more disks, but right now, even if balancing is as perfect as it can be, you will only get 14.4TBs (and it gets exponentially harder the closer you get).  From what I understand, you "safe" limit would be about 11.5TBs (80%). How much do you have stored on the cluster?

![screenshot_2018-11-26_12-53-29](https://user-images.githubusercontent.com/3953102/49041588-7c638080-f17a-11e8-8115-c90f4e797d9a.png)

I am looking around for my re-balance script..  Not sure where I put it (may be lost).

**Comment by BloodBlight** [comment]
OH!  One other "gotcha" just came to mind.  LizardFS (as far as I can tell) tries to but EVERY CHANGE into place at the same time!

So, if you change the goal on a 1GB folder, then it will go through and make just a single part of the EC replication for each block.  THEN starts over at the beginning and does the second, and so on.  SO!  To change goals, you need to capacity for BOTH goals at the same time!  This is what caught me the first time.

This is why I created that script, it forced it to process the goals one file at a time.

So, if you changed the goal of every file at the same time...  Good by space.  :(

Again, that is as for as this USER can tell.

**Comment by hradec** [comment]
You're absolutely right... I did notice that my disk space started to disappear when I changed from cp(2) to ec(6,2)... thats why I kept chaging from ec(6,2) to ec(6,2)... 

Basically, this last 50k chunks stretch still is the last cp(2) chunks being converted to ec(6,2)!!!

Loved the script idea!! I'm definitely writing something like that soon!!

Btw, I tried a master restarted ( as you mentioned earlier ), without any effect... 

But after changing the chunkservers MTU down to 1400 (master still as 1500) and restarting the master again, all of the sudden it started re-balancing a bit faster, and now its actually replicating again! 
![image](https://user-images.githubusercontent.com/806743/49043987-5d1c2180-f181-11e8-8c3f-c855e6e8f6db.png)

My /media folder is down to 1 chunk now!!! YEYY!

And there's lots of  "1 copy to delete" chunks now, going up and down... and everything seems alive again, as you can see in my last 30 minutes: 
![image](https://user-images.githubusercontent.com/806743/49044327-42967800-f182-11e8-934f-d0011b9b7276.png)

It seems there's something going on here with the TCP packets... I've tried jumbo packets before, but never made it work... always got deadlocks and inconsistent connections.

But just now I had this  idea to actually decrease the chunks MTU... maybe it's some overhead going on somewhere... and it seems like it made a HUGE difference! 

Thanks for calculations up there, @BloodBlight... I'll take a careful look at it  tonight... it seems there's a lot on nice interesting info to unpack there!! Really appreciate!!! 

I'll keep you guys informed how things progress... 

Huge thanks man!!

**Comment by hradec** [comment]
another quick update: 

as you can see, now that LFS finally decided to keep replicating, ALL of its replication is going into just ONE disk/chunk: 
![image](https://user-images.githubusercontent.com/806743/49047646-f3097980-f18c-11e8-9e41-6ba5e4c9f271.png)

btw, it's this chunk which thanks to the re-balancing that started after the MTU change and the master restart,  just got back down to 98%.
![deepinscreenshot_select-area_20181126150739](https://user-images.githubusercontent.com/806743/49047717-2b10bc80-f18d-11e8-838d-18f9c1a485ad.png)

But since LFS is choosing that one to fill it up again, I think we're heading to the same "slow down/dead lock" again...

Although, it seems that there's a lot of decreasing happening in all disks as this goes on... so hopefully not! But it doesn't looks like it's re-balancing anymore, since increase is only happening in that one disk/chunk.

For sure there's a bug somewhere, since I didn't had to add new disks or chunks. LFS was able to continue and sort things out aparently, after the MTU and Master restart. 

So there's a sort of "dead lock" state happening from time to time for some reason, and In my opnion has to do with the lack of pre-emptive or even emergencial re-balancing of chunks not happening!

**Comment by hradec** [comment]
just to keep it registered... It seems the MTU 1400 on the chunkservers made a HUGE difference indeed.
![deepinscreenshot_select-area_20181128195537](https://user-images.githubusercontent.com/806743/49198379-acac4a00-f347-11e8-994d-92f7ea57ba13.png)

Data keeps replicating and re-balancing without stopping for the last 2 days! 
![deepinscreenshot_select-area_20181128195710](https://user-images.githubusercontent.com/806743/49198420-d06f9000-f347-11e8-8dc2-f31aef8e95a0.png)

although, /media and /airport1TB are STILL hanging for 1 chunk!!! LOL
at least it's steady and going... hopefully it will finish soon!!!
